### PR TITLE
do not fail for RefExpression

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
ticket to properly support RefExpression: #3024

I tested locally building dnSpy and it reduced errors from 659 to 332

### CODE REVIEW

**Comment by costin-zaharia-sonarsource** [comment]
I think it would be great to add a test in **ExplodedGraphTests** to document actual behavior.

**Comment by costin-zaharia-sonarsource** [comment]
We might want to assert that we have a SymbolicValue for the **Empty** variable after "ref Empty" is processed; but otherwise it LGTM!

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 1]
```diff
@@ -925,26 +959,35 @@ private ExplodedGraphContext(MethodDeclarationSyntax mainMethod, SemanticModel s
...
-                    .DescendantNodes()
-                    .OfType<DeclarationExpressionSyntax>()
-                    .Where(d => d.Designation is SingleVariableDesignationSyntax)
-                    .Select(d => (SingleVariableDesignationSyntax)d.Designation)
-                    .First(d => d.Identifier.Text == identifier);
+                    SymbolType.Variable => this.MainMethod
+                        .DescendantNodes()
+                        .OfType<VariableDeclaratorSyntax>()
+                        .First(d => d.Identifier.ToString() == identifier),
+                    SymbolType.Identifier => (CSharpSyntaxNode) this.MainMethod
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: `CSharpSyntaxNode` is needed to help the compiler figure out the type of the switch expression

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs [hunk 2]
```diff
@@ -925,26 +959,35 @@ private ExplodedGraphContext(MethodDeclarationSyntax mainMethod, SemanticModel s
...
+                        .DescendantNodes()
+                        .OfType<DeclarationExpressionSyntax>()
+                        .Where(d => d.Designation is SingleVariableDesignationSyntax)
+                        .Select(d => (SingleVariableDesignationSyntax)d.Designation)
+                        .First(d => d.Identifier.Text == identifier),
+                };
 
-                return this.SemanticModel.GetDeclaredSymbol(declarationExpression);
+                return this.SemanticModel.GetDeclaredSymbol(expression)
+                    ?? this.SemanticModel.GetSymbolOrCandidateSymbol(expression);
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs]: for some reason `GetDeclaredSymbol` returns null for the `Empty` identifier

**Pull request merged**